### PR TITLE
[WIP][dagster-airbyte] RFC: Prototype `airbyte_assets` asset decorator

### PR DESCRIPTION
**Pull request by maximearmstrong**
## Summary & Motivation

*Note: **The code is still WIP**

### Background

As part of the integration overhaul, we want to explore how we can improve our already existing integration.

Namely, we want to explore how using an approach combining a project, a translator and a decorator would help improving an integration and making it easier to use.

This PR is a RFC and implements a prototype for `dagster-airbyte`, using the a project, a translator and a decorator.

Goals of this RFC:

- Understand frictions in the current dagster-airbyte integration.
  - Does an approach using a project, decorator and translator is better at enabling users than the previous approach?
- Understand whether some patterns makes sense for tools like Airbyte
  - Does the translator pattern makes sense?
  - What about the project pattern?

### Current state

Currently, users can create their Airbyte assets in three different ways in Dagster.

- `load_assets_from_airbyte_instance`
  - Gets all connections from an Airbyte assets and creates Dagster assets for each of them
  - Uses `CacheableAssetsDefinition`
- `load_assets_from_airbyte_project`
  - Gets all connections from an Octavia project directory and local YAML files. Creates Dagster assets for each of the connections.
  - Uses `CacheableAssetsDefinition`
- `build_airbyte_assets`
  - Creates a multi-asset for a single Airbyte connection.
  - Does not use `CacheableAssetsDefinition`, but multi-asset

Looks like this, assuming no overlap in Airbyte connection between approaches:

```python
from dagster import job, EnvVar
from dagster_airbyte import AirbyteResource

my_airbyte_resource = AirbyteResource(
    host=EnvVar("AIRBYTE_HOST"),
    port=EnvVar("AIRBYTE_PORT"),
    # If using basic auth
    username=EnvVar("AIRBYTE_USERNAME"),
    password=EnvVar("AIRBYTE_PASSWORD"),
)

my_connection_assets = with_resources(
    build_airbyte_assets(
        connection_id="87b7fe85-a22c-420e-8d74-b30e7ede77df",
        destination_tables=["releases", "tags", "teams"],
    ),
    {"airbyte": airbyte_instance},
)

my_project_assets  = with_resources(
    [load_assets_from_airbyte_project(project_dir="path/to/airbyte/project")],
    {"airbyte": airbyte_instance},
)

my_instance_assets = load_assets_from_airbyte_instance(
    airbyte = my_airbyte_resource,
)

defs = Definitions(
    assets=[my_connection_assets, my_project_assets, my_instance_assets],
    resources={"airbyte": my_airbyte_resource},
)
```

### Proposition

Use one asset decorator to manage all connections. Use a resource, project and translator in the process.

This proposition
- Prepare the connections in advance, at run time during development and at build time during deployment. Create the assets based on these connections.
- Remove the cacheable assets previously used in `load_assets_from_airbyte_instance` and `load_assets_from_airbyte_project` and adopt a project object with multi-assets
  - like for `DbtProject`, the `AirbyteProject` can be prepared at run time (development) and build time (deployment)

Introduced in this proposition
- `AirbyteProject` object
  - the internal representation of the Airbyte project/configuration.
    - can be created in two ways
      - from_local_configuration(), which takes an Airbyte configuration directory, that was already imported by users with Octavia CLI, see [here](https://airbyte.com/tutorials/version-control-airbyte-configurations).
      - from_resource(), which would import the Airbyte configuration directory using the resource/credentials. (not implemented yet)
  - is used to prepare a connections metadata file, used by the asset decorator.
  - is created based on a single `workspace_id`.
- `airbyte_assets` decorator
  - the decorator that creates all assets for each connection of the airbyte project
  - can `select` and `exclude` connections in the asset decorator (not implemented yet)
- `DagsterAirbyteTranslator` object
  - represent all translation functions required
    - currently represents the connection_to_*_fn passed to `load_assets_from_airbyte_project` and `load_assets_from_airbyte_instance`.
- `materialize_from_sync` method added to the AirbyteResource
  -  it encapsulates the previously existing AirbyteResource.sync_and_poll, and materialize the assets based on the context.
    - though, maybe decoupling the materialization from the resource would be better

```python
from dagster import job, EnvVar
from dagster_airbyte import (
    AirbyteProject, 
    DagsterAirbyteTranslator, 
    AirbyteResource, 
    airbyte_assets
)

my_airbyte_resource = AirbyteResource(
    host=EnvVar("AIRBYTE_HOST"),
    port=EnvVar("AIRBYTE_PORT"),
    # If using basic auth
    username=EnvVar("AIRBYTE_USERNAME"),
    password=EnvVar("AIRBYTE_PASSWORD"),
)

airbyte_project = AirbyteProject.from_local_configuration(
    configuration_dir="path/to/configuration",
    workspace_id="my-workspace-id",
)
airbyte_project.prepare_if_dev()

@airbyte_assets(
    name="my_airbyte_assets",
    project=airbyte_project,
    dagster_airbyte_translator=DagsterAirbyteTranslator(),
)
def my_airbyte_assets(context, airbyte: AirbyteResource):
    # Do something with my resource
    yield from airbyte.materialize_from_sync(context=context)

defs = Definitions(
    assets=[my_airbyte_assets],
    resources={"airbyte": my_airbyte_resource},
)
```

## Request for comments

1. The API
- Any feedback on the API, including the project, decorator, translator and resource is welcome.
  - Does the current implementation make sense in terms of usage?
  - Allowing users to select and exclude connections at the asset decorator level seems like a good idea to make usage more flexible. Other opinions or thoughts?

2. Naming
- `AirbyteProject` vs `AirbyteConfiguration`
  - The proposition includes a AirbyteProject, which respect the `Project` convention that we want to enforce in the integration overhaul.
  - The object represents the configuration more than an actual Airbyte project.
  - Do we still want to name it `AirbyteProject`?

3. Who's responsible for the workspace_id
- In this proposition, the `workspace_id` is passed that the project level, meaning that the project object would represent only one `workspace_id`, if many.
  - Should `workspace_id` be passed in the asset decorator instead, to represent the all the connections of a single workspace?

4. Importing the Airbyte configuration via Octavia CLI for the user
- For this proposition, we need the configuration directory that can be imported via Octavia CLI
  - When using `AirbyteProject.from_local_configuration()`
    - Easy to use to extract the connections and workspaces from the local configuration directory
  - When using `AirbyteProject.from_resource()`
    - Using the resource attributes (credentials), we could import the Airbyte configuration via Octavia CLI
      - We could avoid using the partially initialized resource and the REST API
      - the credentials could be decoupled from the resource to avoid using the resource object here.
      - the configuration directory could then be packaged at build time when deploying, similarly to what we are doing for dbt.
    - Is importing the configuration directory using the credentials a good strategy?

5. Is the configuration directory enough?
- In this proposition, the `AirbyteProject` creates a file when prepared, `connections_metadata.json`, that is what Dagster needs to create the assets.
  - This was created to replicate the cache behiavor previously implemented
  - This is an intermediate step that could be eliminated.
    - The assets could be created using only configuration directory.
      - No need to "prepare" the Airbyte project created with `AirbyteProject.from_local_configuration()`
      - We would still need to either import the configuration directory via Octavia CLI or fetch the connections via the REST API when using `AirbyteProject.from_resource()`.
    - Any objection with removing that intermediate step?

## Additional thoughts

1. One of the initial propositions for this prototype was to only refactor the current `build_airbyte_assets` in an asset decorator. This approach should be dropped.
  - In this approach, we would keep `load_assets_from_airbyte_instance` and `load_assets_from_airbyte_project` as-is, which does not make much sense in hindsight.
  - the asset decorator was initially `airbyte_assets` , now named `airbyte_single_connection_asset` in this PR. A method was added to the resource as well, initially `materialize_from_sync`, now named `materialize_single_connection_from_sync`.
    - I'm just waiting for the first round of feedback to confirm the main proposition before removing this.

## To-do

- [x] Implement the translator to replace the function passed to `load_assets_from_airbyte_instance` and `load_assets_from_airbyte_project`
- [x] Implement a `AirbyteProject.from_local_configuration()` to reflect the project and connections 
- [x] Implement a `AirbyteProject.from_resource()` to reflect the project and connections 
- [ ] Update code to use `AssetSpec` instead of `AssetOut`
- [ ] Update code to `select` and `exclude` connections in the asset decorator.
- [ ] Update and add tests

### CODE REVIEW

**Comment by maximearmstrong** [comment]
* **#22847** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22847?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @maximearmstrong and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by sryza — COMMENTED** [review]
Do we have a sense of how often our different Airbyte APIs are used? My understanding was that:
- Octavia / `load_assets_from_airbyte_project` are barely used, because Octavia is barely used. We could probably deprecate this.
- `load_assets_from_airbyte_instance` is widely used

> AirbyteProject vs AirbyteConfiguration

I think the name `DbtProject` makes sense because dbt itself has a "project" concept. If Airbyte doesn't have a similar project concept (outside of Octavia), I think "configuration" makes more sense.

Relatedly, given that Airbyte doesn't seem to have an analogous compilation step to dbt, it seems like we don't need `prepare_if_dev`?

**Comment by maximearmstrong** [comment]
Thanks @sryza, that's good context.

I agree that `AirbyteConfiguration` is closer to what we want to represent here.

To create the assets, we need the connections and their metadata. We can get these by either by using the REST API or the configuration directory provided by Octavia CLI.

When using something like `AirbyteConfiguration.from_resource()`, my first thought was to use Octavia CLI and manage the object similarly to what we do for dbt:
- during development, import the configuration directory at run time when `dagster dev` is invoked;
- during deployment, import the configuration direction once at build time.

This would avoid using a partially initialized resource and the REST API to fetch the connection IDs during run time in production.

Do you think deprecating the Octavia approach in the public API, but using it under the hood, could still make sense? Or should we avoid using that approach?

**Comment by sryza** [comment]
Do we know if the Octavia CLI is stable? I couldn't find much information on it in their docs. I'm a little worried they'll end up abandoning it.

**Comment by sryza** [comment]
[Here](https://airbyte.com/tutorials/version-control-airbyte-configurations), they say:

> Note: This tutorial leverages Octavia CLI, which is an alpha unofficial CLI that won't be maintained. Since the publication of this tutorial, Airbyte has released an official [Terraform Provider](https://registry.terraform.io/providers/airbytehq/airbyte/0.1.0), which we would advise to use instead of the CLI.

**Comment by maximearmstrong** [comment]
> [Here](https://airbyte.com/tutorials/version-control-airbyte-configurations), they say:
> 
> > Note: This tutorial leverages Octavia CLI, which is an alpha unofficial CLI that won't be maintained. Since the publication of this tutorial, Airbyte has released an official [Terraform Provider](https://registry.terraform.io/providers/airbytehq/airbyte/0.1.0), which we would advise to use instead of the CLI.

Good catch @sryza - I'd missed that.

Then, using the REST API to fetch the connections would make more sense. 

I'd still be in favor of using a process where the connections are fetched at run time during development, and at build time only during deployment.

**Comment by sryza** [comment]
> I'd still be in favor of using a process where the connections are fetched at run time during development, and at build time only during deployment.

When you say "at build time only during deployment", are you able to say a little more about how you envision that working?

File: python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_decorator.py [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+    group_name: Optional[str] = None,
+    io_manager_key: Optional[str] = None,
+    normalization_tables: Optional[Mapping[str, Set[str]]] = None,
+    deps: Optional[Iterable[Union[CoercibleToAssetKey, AssetsDefinition, SourceAsset]]] = None,
+    upstream_assets: Optional[Set[AssetKey]] = None,
+    schema_by_table_name: Optional[Mapping[str, TableSchema]] = None,
+    freshness_policy: Optional[FreshnessPolicy] = None,
+    stream_to_asset_map: Optional[Mapping[str, str]] = None,
+    auto_materialize_policy: Optional[AutoMaterializePolicy] = None,
+    dagster_airbyte_translator: Optional[DagsterAirbyteTranslator] = None,
```
Comment (schrockn) [inline python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_decorator.py:58]: would prefer to use a single `AssetSpec` here rather than exploding out all the parameters

Comment (maximearmstrong) [reply python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_decorator.py:58]: I agree completely - the `airbyte_single_connection_asset` asset decorator was the initial test, derived from the current `build_airbyte_assets`. 

Most likely, this approach will not be adopted, in favour of the `airbyte_assets` asset decorator, derived from `load_assets_from_airbyte_resource`. This asset will represent all connections in Airbyte and allow users to select which connections to materialize, similar to what we do with `dbt_assets`. The plan is to use `AssetSpec` for `airbyte_assets`.

**Comment by sryza** [comment]
> When you say "at build time only during deployment", are you able to say a little more about how you envision that working?

Oops I thought this piece depended on Octavia. Now that I realize it doesn't, no need to answer this question.

A downside of the "fetch from Airbyte at CI time" approach, relative to `CacheableAssetsDefinition`, is that the user needs to re-run their CI / rebuild their Docker container in order for changes made in Airbyte to get reflected in Dagster.  Vs. with `CacheableAssetsDefinition`, all they need to do is hit the "Reload definitions" button or GraphQL endpoint.

The benefits of avoiding `CacheableAssetsDefinition` are high, so might be worth this downside, but I think it's worth trying to find some users who rely on the current approach and understanding how this would affect them.

**Comment by maximearmstrong** [comment]
> The benefits of avoiding CacheableAssetsDefinition are high, so might be worth this downside, but I think it's worth trying to find some users who rely on the current approach and understanding how this would affect them.

@sryza that's a good point. I was thinking about something similar to what @schrockn describes here:

> At the end of this I'd like this to be a pattern we can copy for other cases similar to this, so developing a common terminlogy would be good. E.g. should be call connections_metadata.json an "auto manifest" or something since it services the purpose that a dbt manifest serves and we automatically produce it.

We could create the `connections_metadata.json` file and use it to create the assets - at run time during development, at build time during deployment.

That said, the two main things we need to clarify are:
- How can I avoid problems if the REST api fails during deployment?
- How would redeployment to refresh connections affect users?

**Comment by maximearmstrong** [comment]
> How can I avoid problems if the REST api fails during deployment?

My current thoughts on how to handle the connections are:

- Use an `AirbyteMetadata` object
  - Can be instantiated by specifying the instance and the output directory. The instance can be an `AirbyteResource`.
  
    ```python
    airbyte_metadata = AirbyteMetadata(instance=airbyte_resource, output_directory=output_directory)
    ```
  - For local development, can be prepared at run time. Ideally, we would decouple the creation of the metadata file from the constructor.
    
    ```python
    airbyte_metadata.prepare_if_dev()
    ```
  - For deployment, can be prepared are build time using dagster-airbyte CLI.
  
    ```bash
    dagster-airbyte metadata prepare-and-package
    ```
    - By default, we could emit a warning if the REST api fails, instead of raising an error. Fallback options:
      1. Use a `connections_metadata.json` file that was committed in the repo.
      2. Creates an empty list in `connections_metadata.json`, to prevent deployment from failing

    - If users want the deployment to fail if the REST api fails, use flag. 
    
      ```
      `dagster-airbyte metadata prepare-and-package  --raise-on-airbyte-api-failure`
      ```
    - If users want to retry if the REST api fails, use flag. 
    
      ```
      `dagster-airbyte metadata prepare-and-package  --retry-on-airbyte-api-failure`
      ```
 
The thing I'm most unsure about at the moment is what strategy to adopt if the REST api fails. This is also true for the current `load_assets_from_airbyte_instance` because it calls the api when creating the assets. Any other ideas are welcome!

> How would redeployment to refresh connections affect users?

@sryza are we comfortable releasing an experimental MVP and test it with users?

> Side note: I'm not a big fan of the "Translator" pattern and am interested in an alternative.

@schrockn similar to the question above, could we use the Translator pattern for an experimental MVP and work on alternatives? The Translator pattern is used in dagster-dbt, so it might be interesting to have a broader discussion on the subject.

For the timeline, we would like to have an initial experimental MVP by the end of next week.

**Comment by sryza** [comment]
So at a high level we're considering two alternatives here, if I understand correctly:
1. Fetch from Airbyte at the time we load definitions (current approach)
2. Fetch from Airbyte at CI time

I think I still lean towards (1), especially given this approach we recently devised for simplifying how this pattern is handled within the framework: https://github.com/dagster-io/internal/discussions/10666.

The main reason I lean towards (1) is that it gives responsibilities to lifecycle stages that match my expectations for them:
  - CI, which is typically triggered in response to changes to a git repository, is solely responsible for compiling, packaging, and publishing the contents of that repository.
  - The "load definitions" operation is responsible for syncing a Dagster deployment's state with an external source of truth on definitions. That external source of truth can the contents of a git repository, but it can also be an external service.

Whether or not we end up going with (1), I think we should try to reach alignment on one of these options before shipping a new API.

File: python_modules/libraries/dagster-airbyte/dagster_airbyte/dagster_airbyte_translator.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
+from dagster import AssetKey
+
+from dagster_airbyte.asset_defs import AirbyteConnectionMetadata
+
+
+class DagsterAirbyteTranslator:
```
Comment (OwenKephart) [inline python_modules/libraries/dagster-airbyte/dagster_airbyte/dagster_airbyte_translator.py:6]: Peanut gallery checking in, but if we're making a concerted effort here to improve our suite of integrations, I think now would be an excellent time to do a bit of cross-integration refactoring / abstraction.

The most glaring thing to me is that `DagsterAirbyteTranslator` and `DagsterDbtTranslator` should really share a common base class (and probably "think" in terms of AssetSpecs, even if we retain the exploded methods for backcompat).

i.e. there should be a generic:

```python
class DagsterTranslator(Generic[T]):
    
    def get_asset_spec(identifier: str, metadata: T) -> AssetSpec:
        ...
```

or something of the sort, which all of our integrations can inherit from. Then `DagsterAirbyteTranslator` would look like:

```python
class DagsterAirbyteTranslator(DagsterTranslator[AirbyteConnectionMetadata]):
    ...
```

and users could override things like:

```python
class MyCustomTranslator(DagsterAirbyteTranslator):

    def get_asset_spec(identifier: str, metadata: AirbyteConnectionMetadata) -> AssetSpec:
         return super().get_asset_spec(identifier, metadata)._replace(
              <stuff i want to replace>
         )
```

This a) reduces boilerplate and b) means that we don't need to update every single one of these things if we ever decide to add something to AssetSpec. I also thing it's probably generally more convenient for a user to write a single function that sets multiple properties of a spec rather than one function per property.

So that's the Translator side of the world, but I also think there's likely an opportunity to make a generic "integration assets decorator", or at least simplify the process a bunch -- assuming we get to the point where we can just make a simple function that returns a list of AssetSpecs (with the help of this Translator object), then we can probably create some generic function which takes in:

- domain specific function to get some raw blob of metadata (e.g. for dbt this would be the code to get the manifest.json file)
- a translator object to turn that into specs

and then combines those together to create a decorator. this bit is probably less pressing than the Translator standardization piece, mostly just tossing it out there as an idea.

TL;DR: I really think we should standardize these things sooner rather than later to make future integration work faster / easier / more consistent

**Comment by maximearmstrong** [comment]
> Whether or not we end up going with (1), I think we should try to reach alignment on one of these options before shipping a new API.

@sandy That makes sense. After discussing with @yuhan, we'll wait to see the conclusion of the Cacheable Assets discussion before revisiting this prototype, so that we have more clarity on the strategies we can adopt here.

In the meantime, we will work on a prototype/RFC for the translator pattern, to address @schrockn and @OwenKephart's feedback.

**Review by cmpadden — COMMENTED** [review]
Hi @maximearmstrong , would it be possible to convert this to a DRAFT until ready for review? Thanks!

**Comment by maximearmstrong** [comment]
Closing this RFC - asset decorator implemented in https://github.com/dagster-io/dagster/pull/26432.
